### PR TITLE
feat(message-input): dt-1092 message input vue 3

### PR DESCRIPTION
**Pull request by sanjeevdialpad**
# Message Input Recipe

## :hammer_and_wrench: Type Of Change

- [ ] Fix
- [x] Feature
- [ ] Refactoring
- [ ] Documentation

## :book: Description

Message Input Recipe

## :pencil: Checklist

- [ ] I have reviewed my changes
- [ ] I have added tests
- [ ] I have added all relevant documentation
- [ ] I have validated components with a screen reader
- [ ] I have validated components keyboard navigation
- [ ] I have considered the performance impact of my change
- [ ] I have checked that my change did not significantly increase bundle size
- [ ] I am exporting any new components or constants in the index.js in the component directory
- [ ] I am exporting any new components or constants in the index.js in the root

### CODE REVIEW

**Comment by sanjeevdialpad** [comment]
This is the first pass or a more basic version of message input.

Still to be implemented as separate PR:
1. Character limit
2. Error banners

**Review by iropolo — DISMISSED** [review]
Nice job here!, looks awesome, I will point a few visually catches before go through the code but it is great initial!


Emoji picker------/////


1. Emoji picker has some margins
This is added by the class `d-p16` in the `dt-popover-content` element. Im not sure if it should be there because it interfered the emoji-picker style.

Screenshot 2023-06-12 at 11 31 55

3. Emoji picker doesn't update the skin tone color.

4. When I select an emoji it doesnt appear correctly in the input field

image

5. Emoji picker input search should be focussed on mounted, this is already in place in the component but I don't see it working here, hmm... maybe I should expose a focus input search method so you can trigger it after open the emoji picker... but take a look if theres is something you can do on your side because the component already has a focus search input on mounted.

6. If I press enter on empty input search it throw an error:

image


7. If I press enter in a tabset it close the emoji picker (it is not working as how the component should)

image

8. When I reach the last tabset in the emoji selector it doesn't jump to the skin selector, its jumping directly to the tabs (it is not working as how the component should)

9. The emoji picker has a color issue on the fixed label side 

Screenshot 2023-06-12 at 11 42 40

```
.d-popover__dialog {
    --popover-color-background: var(--bgc-secondary);
```

Seems like is the popover which is adding this things, I guess you can place a white background color to override these in the `dt-popover-content` element.

10. The input search persist the previous search, it should be empty.

11. The emoji picker trough an error if you press enter in a no result emoji list.

image

image

**Comment by sanjeevdialpad** [comment]
@iropolo thank you for letting me know of these errors. this is very helpful.

The emoji picker specific errors excluding padding are related to the component which isnt part of this component/Recipe. I wonder if i should open bugs for it separately. @braddialpad 

I will go through each of the errors reported by you and create tickets or solve it if possible in this PR itself.

**Comment by iropolo** [comment]
> @iropolo thank you for letting me know of these errors. this is very helpful.
> 
> The emoji picker specific errors excluding padding are related to the component which isnt part of this component/Recipe. I wonder if i should open bugs for it separately. @braddialpad
> 
> I will go through each of the errors reported by you and create tickets or solve it if possible in this PR itself.

Lets check this in parts....

3 )
If you add this in the `<dt-emoji-picker`
```
:skin-tone="skinTone"
@skin-tone="skinTone = $event"
```

and this inside `data () {`

`skinTone: 'Default',`

You fixed the skin tone change issue...

5-7-8)

Lets see now regarding the keyboard accesibility.
This is being affected by tippy somehow, because if you extract the `<dt-emoji-picker` from the `<template #content>` and put it outside, somewhere with a `v-if="emojiPickerOpened"` you will see the component mounts as expect and without problems, the input search is focused and the navigation with keyboard is working.


6-11)
Regarding the error on empty or `No result` emoji after press Enter it is because this method is not validating a valid emoji:

```
onSelectEmoji (emoji) {
      this.inputValue = this.inputValue + emoji.shortname;
      this.emojiPickerOpened = false;
    },
```

It can be simply fixed adding this `if (!emoji) { return; }` or sth..

**Comment by sanjeevdialpad** [comment]
@iropolo ty. The accessibility issue i was still figuring out. Your solution fixes it.

On a unrelated note, all the labels currently are hardcoded on this recipe, should we pass it down from props since we need product to do i18n ?

**Comment by iropolo** [comment]
> 

Certainly yes, this should not be hardcoded in the final product.
All labels should be managed by props to give the possibility to i18n to write it.

**Comment by sanjeevdialpad** [comment]
https://dialpad.atlassian.net/browse/DIALTONE-1094


Character Limit code has been added for the recipe.

**Comment by sanjeevdialpad** [comment]
Creating separate PRs for characterLimit and error message.

Character Limit logic: https://github.com/dialpad/dialtone-vue/pull/1019

**Comment by iropolo** [comment]
Created #1020 to unblock this PR and continue working.

File: recipes/conversation_view/message_input/message_input.mdx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { Canvas, Story, Subtitle, Controls, Meta } from '@storybook/blocks';
+import * as DtRecipeMessageInputStories from './message_input.stories.js';
+
+<Meta of={DtRecipeMessageInputStories}/>
+
+# Message Input Recipe
+
+<Subtitle>
+  Message Input recipe wraps the Rich Text Editor to be used for conversations.
```
Comment (braddialpad) [inline recipes/conversation_view/message_input/message_input.mdx]: In theory doesn't have to be used only for conversations. ex: writing notes or something

File: recipes/conversation_view/message_input/message_input.vue [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+    searchResultsLabel: {
+      type: String,
+      default: 'Search results',
+    },
+
+    searchPlaceholderLabel: {
+      type: String,
+      default: 'Search...',
+    },
```
Comment (braddialpad) [inline recipes/conversation_view/message_input/message_input.vue]: add emoji in front of the name of each of these, it is not clear in the docs that they are for the emoji picker portion of this component

ex: `searchPlaceholderLabel` = `emojiSearchPlaceholderLabel`

File: recipes/conversation_view/message_input/message_input_default.story.vue [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      @submit="$attrs.onSubmit"
+      @focus="$attrs.onFocus"
+      @blur="$attrs.onBlur"
+      @input="$attrs.onInput"
+    >
+      <template
+        #footerLeft
+      >
+        <div class="message-input-recipe-subtitle d-fc-black-500">
+          <b>Dwight</b> is typing...
```
Comment (braddialpad) [inline recipes/conversation_view/message_input/message_input_default.story.vue:33]: lol I love how you always use dwight for examples.

**Review by braddialpad — COMMENTED** [review]
Just adding some doc stuff since @iropolo has already reviewed thoroughly.

Looks great!

File: recipes/conversation_view/message_input/message_input.vue (hunk 2 above)
Comment (sanjeevdialpad) [reply recipes/conversation_view/message_input/message_input.vue]: changed.

File: recipes/conversation_view/message_input/message_input_default.story.vue (hunk 3 above)
Comment (sanjeevdialpad) [reply recipes/conversation_view/message_input/message_input_default.story.vue:33]: Not just notes, wait till you see DPM beta datastore. 😛

**Comment by sanjeevdialpad** [comment]
@iropolo Emoji pickers been fixed.

**Review by braddialpad — APPROVED** [review]
Looks good by me, you just need to fix the accessibility tests.

File: recipes/conversation_view/message_input/message_input.vue [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+      <slot name="middle" />
+      <!-- Section for the bottom UI -->
+      <section class="d-d-flex d-jc-space-between d-mx8 d-my4">
+        <!-- Left content -->
+        <div class="d-d-flex">
+          <dt-button
+            size="sm"
+            circle
+            importance="clear"
+          >
```
Comment (braddialpad) [inline recipes/conversation_view/message_input/message_input.vue]: These buttons need aria-label, will have to be a prop because of i18n

**Comment by braddialpad** [comment]
Just noticed there is also an error occurring on this page.

```
button.vue:314 You cannot have a circle button with kind: default and importance: primary as it does not exist in our design system.
```

File: recipes/conversation_view/message_input/message_input.vue (hunk 4 above)
Comment (sanjeevdialpad) [reply recipes/conversation_view/message_input/message_input.vue]: got it.

**Comment by sanjeevdialpad** [comment]
Will be making the error banner and the image selector in a different PR. fixed all issues.

**Review by braddialpad — APPROVED** [review]

**Pull request merged**